### PR TITLE
Render user actions separately

### LINKED ISSUE #577: Display actions (such as join/leave/nick change) differently
It's confusing to display them same as normal messages.

Here's how it's done in HexChat for example:
hexchat_xpaw_ _holmes_freenode__ steamdb_2016-01-11_13-39-21

### PR DESCRIPTION
**Pull request by xPaw**
This fixes #577, #404.

Needs to be done before merging this:
- [x] Merge #596
- [x] Merge #597 
- [x] Test and fix and username/action colors if they're terrible
- [x] Look into using font awesome icons for actions

### CODE REVIEW

**Comment by dgw** [comment]
What I can see of the previews from my phone looks good. Perfect or not, this is a big step forward from how it looks now. :+1:

**Comment by dgw** [comment]
See #571, though. The user `button` elements were changed to `a` to allow copying the text in Firefox. The new tpl should use `a`s too.

**Comment by dgw** [comment]
Never mind, it moved to #574 and still needs review. But the change will probably be merged.

**Comment by xPaw** [comment]
I've tweaked the action texts, and it now looks like this:

yoloswag_ _shout_-_google_chrome_2016-01-17_22-22-33

**Comment by xPaw** [comment]
Also formatted /me actions and added "Notice" prefix to notices.

wowtest_ _shout_-_google_chrome_2016-01-17_22-45-14

**Comment by MaxLeiter** [comment]
+1 looks awesome

File: client/css/style.css [hunk 1]
```diff
@@ -603,11 +603,6 @@ button {
 #chat .msg:last-child .text {
 	padding-bottom: 10px;
 }
-#chat .msg .type {
-	color: #ccc;
-	display: none;
-	font-style: normal;
-}
```
Comment (astorije) [inline client/css/style.css]: @xPaw, what does that remove already?

File: client/css/style.css [hunk 2]
```diff
@@ -657,16 +651,6 @@ button {
...
-#chat .join .type,
-#chat .part .type,
-#chat .mode .type,
-#chat .nick .type,
-#chat .kick .type,
-#chat .quit .type,
-#chat .quit .type,
-#chat .topic .type {
-	display: inline;
-}
```
Comment (astorije) [inline client/css/style.css]: Why this change?

File: client/css/style.css [hunk 3]
```diff
@@ -681,19 +665,38 @@ button {
...
-	content: "* ";
+#chat .action .from:before {
+	content: "★";
+	color: #f39c12;
 }
-#chat .action .user:after {
-	content: "";
+#chat .notice,
+#chat .chan .notice .user {
+	color: #0074D9 !important;
```
Comment (astorije) [inline client/css/style.css]: What happens if `!important` is absent? Can we easily change something to remove its need? Not that I would hold this PR for this, but really we should get rid of these bastards :-)

File: client/css/style.css [hunk 4]
```diff
@@ -681,19 +665,38 @@ button {
...
+	color: #2ECC40;
+}
+#chat .action .text,
 #chat .chan .action .user {
 	color: #f39c12 !important;
 }
-#chat .action .user:before {
-	content: "* ";
+#chat .action .from:before {
+	content: "★";
```
Comment (astorije) [inline client/css/style.css]: Just a random thought, what about using Font Awesome (as we do everywhere else) for these icon-like goodies?

File: client/css/style.css [hunk 5]
```diff
@@ -681,19 +665,38 @@ button {
...
+	color: #f39c12;
 }
-#chat .action .user:after {
-	content: "";
+#chat .notice,
+#chat .chan .notice .user {
+	color: #0074D9 !important;
 }
+#chat .notice .user:before {
+	content: "Notice: ";
```
Comment (astorije) [inline client/css/style.css]: Stupid me, but how do you trigger these? I played with a few commands and stayed on channels around, but it didn't show up.

File: client/js/shout.js [hunk 6]
```diff
@@ -183,14 +183,42 @@ $(function() {
...
+			"action",
+		].indexOf(type) !== -1) {
+			switch (type) {
+			case "join": data.msg.formattedAction = "has joined the channel"; break;
+			case "mode": data.msg.formattedAction = "sets mode"; break;
+			case "kick": data.msg.formattedAction = "has kicked"; break;
+			case "nick": data.msg.formattedAction = "is now known as"; break;
+			case "part": data.msg.formattedAction = "has left the channel"; break;
+			case "quit": data.msg.formattedAction = "has quit"; break;
+			case "topic": data.msg.formattedAction = "has changed the topic to:"; break;
```
Comment (astorije) [inline client/js/shout.js]: From my experience, topics are often long, full of special characters (colons, dashes, ...), like:

screen shot 2016-01-19 at 23 59 06

What about displaying the new topic on its own line, like the following:

screen shot 2016-01-20 at 00 03 07

Actually, changing topic is often a big deal that needs to be seen from everyone: I would display it differently to catch the eye, like in bold, or italic, ... What do you think?

File: client/views/msg_action.tpl [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+<div class="msg {{type}} {{#if self}}self{{/if}}">
+	<span class="time">
+		{{tz time}}
+	</span>
+	<span class="from"></span>
+	<span class="text">
+		<a href="#" class="user" style="color: #{{stringcolor from}}">{{mode}}{{from}}</a> {{formattedAction}} {{{parse text}}}
+	</span>
+</div>
+{{/each}}
```
Comment (astorije) [inline client/views/msg_action.tpl]: What do you think we should do for the generated templates here (I'm talking specifically here, not long term solution for handling assets)? Adding the `grunt`ed files once this is ready to be merged?

**Comment by astorije** [comment]
@xPaw, I left you a bunch of minor comments inline. This looks pretty awesome, really!
I hope we can merge that soon :-)

File: client/css/style.css (hunk 1 above)
Comment (xPaw) [inline client/css/style.css]: As you can see in the `msg` template, `<em class="type">{{type}}</em>` was removed as it's not needed anymore, it was used to display the action.

File: client/css/style.css (hunk 2 above)
Comment (xPaw) [inline client/css/style.css]: Same as above.

File: client/css/style.css (hunk 3 above)
Comment (xPaw) [inline client/css/style.css]: I tried it, but nickname wasn't colored correctly, I think due to inline nick colors.

File: client/css/style.css (hunk 5 above)
Comment (xPaw) [inline client/css/style.css]: Use `/notice`, but shout doesn't display your own sent notices.

Comment (dgw) [inline client/css/style.css]: Easiest way is probably to do `/msg NickServ <something>`, as services on most popular networks usually reply in notices.

**Comment by xPaw** [comment]
I've added displaying kick/part/quit messages, which will fix #404. However for this to work, `slate-irc` needs to be updated to the latest version, so that the message is passed through correctly.

File: client/js/shout.js (hunk 6 above)
Comment (MaxLeiter) [inline client/js/shout.js]: +1; would like the title to have its own like and be easy to see (although maybe a different color instead of bold/italics?

File: client/css/style.css (hunk 4 above)
Comment (xPaw) [inline client/css/style.css]: Not a terrible idea, but what kind of icons would fit for this purpose? Each different action of course have it's own icon (kicks, topic changes, etc).

File: client/js/shout.js (hunk 6 above)
Comment (xPaw) [inline client/js/shout.js]: Doing styling is not a good idea as a topic on it's own can contain irc bold/italic/colors.

**Comment by xPaw** [comment]
@astorije I've added font awesome icons, looks pretty cool.

File: client/css/style.css (hunk 5 above)
Comment (astorije) [inline client/css/style.css]: Thanks @dgw, I'll try that.
@xPaw, I guess that's why you opened #590 then :-)

File: client/js/shout.js (hunk 6 above)
Comment (astorije) [inline client/js/shout.js]: What about just the new topic on its own line then?

**Comment by astorije** [comment]
> EDIT: Made topic changes green:

Wheren't you against bold/italic/colors in topic line? I'm OK with anything at the moment really.

File: client/css/style.css [hunk 8]
```diff
@@ -681,19 +665,58 @@ button {
...
-#chat .notice .type {
-	display: none;
+#chat .join .from:before {
+    font-family: FontAwesome;
+	content: "\f090";
+	color: #2ECC40;
+}
+#chat .kick .from:before {
+    font-family: FontAwesome;
+	content: "\f00d";
```
Comment (astorije) [inline client/css/style.css]: For `/kick`, I'm not too fond of this icon, usually used for closing something and sometimes for deleting (but less and less, I notice).
I think I'd prefer `ban` (https://fortawesome.github.io/Font-Awesome/icon/ban/). If really you don't like it, the same icon as leave would work as the consequences are the same (the user leaves the channel) although I still prefer `ban` to that.
Alternatives could be https://fortawesome.github.io/Font-Awesome/icon/exclamation-triangle/, https://fortawesome.github.io/Font-Awesome/icon/exclamation-circle/, https://fortawesome.github.io/Font-Awesome/icon/exclamation/ (but these I would prefer to notify a warning within the app) and if really we want to go esoteric, these could do: https://fortawesome.github.io/Font-Awesome/icon/bolt/, https://fortawesome.github.io/Font-Awesome/icon/crosshairs/, https://fortawesome.github.io/Font-Awesome/icon/crosshairs/, https://fortawesome.github.io/Font-Awesome/icon/thumbs-o-down/, https://fortawesome.github.io/Font-Awesome/icon/frown-o/, https://fortawesome.github.io/Font-Awesome/icon/bomb/.
Meh, `ban` still conveys the most meaning, IMO :-)

File: client/css/style.css [hunk 9]
```diff
@@ -681,19 +665,58 @@ button {
...
+	color: #2ECC40 !important;
+}
+#chat .topic .from:before {
+    font-family: FontAwesome;
+	content: "\f0a1";
+	color: #2ECC40;
+}
+#chat .mode .from:before {
+    font-family: FontAwesome;
+	content: "\f013";
```
Comment (astorije) [inline client/css/style.css]: This one is tricky... But we can't use the `cog` here as it's usually used for settings, and we actually use it for that reason in the app.
Possible alternatives:
- https://fortawesome.github.io/Font-Awesome/icon/info-circle/
- https://fortawesome.github.io/Font-Awesome/icon/key/
- https://fortawesome.github.io/Font-Awesome/icon/terminal/
- https://fortawesome.github.io/Font-Awesome/icon/info/

File: client/css/style.css [hunk 10]
```diff
@@ -681,19 +665,58 @@ button {
...
+#chat .nick .from:before {
+    font-family: FontAwesome;
+	content: "\f007";
+	color: #2ECC40;
 }
-#chat .action .user:before {
-	content: "* ";
+#chat .action .text,
+#chat .action .user {
+	color: #F39C12 !important;
```
Comment (astorije) [inline client/css/style.css]: Why this?

File: client/css/style.css (hunk 5 above)
Comment (astorije) [inline client/css/style.css]: @xPaw, can we find a Font Awesome icon for this, for consistency with the other actions? Maybe the same as `mode`?
Also, can you post a screenshot of this please? :)

**Comment by astorije** [comment]
These do look pretty cool... I look forward to that! :-)

**Comment by xPaw** [comment]
Changed icons as per feedback

chrome_2016-01-24_12-01-16

> Wheren't you against bold/italic/colors in topic line? I'm OK with anything at the moment really.

You are right, I've removed green again. And I would argue it's not the best idea to make it a separate line too.

**Comment by astorije** [comment]
I don't think this should have been merged already, while the discussion was ongoing...
